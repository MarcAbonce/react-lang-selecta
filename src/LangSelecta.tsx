import React from 'react';

import _localeRegions from './data/locale_regions.json';

type LocaleRegions = {
  [locale: string]: {
    name: string;
    regions: string[];
  }
};

const localeRegions: LocaleRegions = _localeRegions;

type LangSelectaProps = {
  langs: string[];
};

function getFlagEmoji(regionCode: string) {
  const codePoints = regionCode.split('').map((char) => 127397 + char.charCodeAt(0));
  return String.fromCodePoint(...codePoints);
}

function pickRandomFlagEmoji(lang: string) {
  if (lang in localeRegions) {
    const regions = localeRegions[lang].regions;
    const index = Math.floor(Math.random() * regions.length);
    return getFlagEmoji(regions[index]);
  } else {
    // return United Nations flag as a fallback
    return "ðŸ‡ºðŸ‡³";
  }
}

function LangSelecta({ langs } : LangSelectaProps) {
  const langsData = langs.map(lang => ({
    code: lang,
    name: lang in localeRegions ? localeRegions[lang].name : lang,
    flag: pickRandomFlagEmoji(lang),
  }));

  return (
    <select>
      {langsData.map(lang => (
        <option value={lang.code}>{lang.flag} {lang.name}</option>
      ))}
    </select>
  );
}

export default LangSelecta;
