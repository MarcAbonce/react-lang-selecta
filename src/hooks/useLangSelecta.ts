/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { useMemo } from 'react'
import leftPad from 'left-pad'

import { localeRegions } from '../data'
import { getFlagEmoji } from '../utils/getFlagEmoji'
import { getRandomListIndex } from '../utils/getRandomListIndex'
import { shuffleList } from '../utils/shuffleList'

export const pickRandomFlagEmoji = (lang: string): string => {
  if (lang in localeRegions) {
    const regions = localeRegions[lang].regions
    const index = getRandomListIndex(regions.length)
    return getFlagEmoji(regions[index])
  } else {
    // return United Nations flag as a fallback
    return 'ðŸ‡ºðŸ‡³'
  }
}

export interface LangSelectaHookProps {
  langs: string[]
}

export interface LangSelectaHookResult {
  langsData: Array<{
    code: string
    name: string
    flag: string
    optionName: string
  }>
}

export const useLangSelecta = ({ langs }: LangSelectaHookProps): LangSelectaHookResult => {
  const langsData = useMemo(() => shuffleList(langs).map(lang => {
    const langCode = lang.split('-')[0]
    const flag = pickRandomFlagEmoji(langCode)
    const name = langCode in localeRegions ? localeRegions[langCode].name : lang
    const optionName = flag + leftPad(name, name.length + 1)

    return {
      code: langCode,
      name,
      flag,
      optionName
    }
  }), [langs])

  return { langsData }
}
