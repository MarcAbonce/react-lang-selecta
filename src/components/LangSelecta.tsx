/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { forwardRef, type ComponentPropsWithoutRef } from 'react'

import { useLangSelecta } from '../hooks/useLangSelecta'
import type { ExtraLangData } from '../types'

export type LangSelectaProps = {
  langs: string[]
  extraLangData?: ExtraLangData
} & ComponentPropsWithoutRef<'select'>

const LangSelecta = forwardRef<HTMLSelectElement, LangSelectaProps>(
  ({ langs, extraLangData, ...rest }, ref): JSX.Element => {
    const { langsData } = useLangSelecta({ langs, extraLangData })

    return (
      <select ref={ref} {...rest}>
        {langsData.map(lang => (
          <option key={lang.code} value={lang.code}>{lang.optionName}</option>
        ))}
      </select>
    )
  }
)

LangSelecta.displayName = 'LangSelecta'

export { LangSelecta }
