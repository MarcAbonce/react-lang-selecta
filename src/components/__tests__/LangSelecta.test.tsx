/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { render } from '@testing-library/react'

import { useLangSelecta } from '../../hooks/useLangSelecta'

import { LangSelecta } from '../LangSelecta'

jest.mock('../../hooks/useLangSelecta')

const mockUseLangSelecta = jest.mocked(useLangSelecta)

describe('LangSelecta component', () => {
  beforeEach(() => {
    mockUseLangSelecta.mockReturnValue({
      langsData: [
        {
          code: 'yi',
          name: 'Yugopotamian',
          flag: '🏴‍☠️'
        },
        {
          code: 'pl',
          name: 'Pig Latin',
          flag: '🇻🇦'
        }
      ]
    })
  })

  it('does not change', () => {
    expect(render(<LangSelecta langs={['yi', 'pl']} />).container).toMatchSnapshot()
  })
})
