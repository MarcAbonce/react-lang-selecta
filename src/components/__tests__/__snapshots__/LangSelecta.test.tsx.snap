// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LangSelecta component does not change 1`] = `
<div>
  <select>
    <option
      value="yi"
    >
      🏴‍☠ Yugopotamian
    </option>
    <option
      value="pl"
    >
      🇻🇦 Pig Latin
    </option>
  </select>
</div>
`;
